#Spam detector

* Training datasets of ham and spam can be found at [Apache SpamAssassin's public datasets](https://spamassassin.apache.org/old/publiccorpus/)

* ML paradigm  used was Offline learning

* Problem type: Supervised Learning

* Problem type:  Classification task



# Step 1 : Prepare the workspace

In [1]:
#Create the folders that will be filled with examples to train the ML model.
!mkdir /content/ham_examples
!mkdir /content/spam_examples

In [2]:
#Decompress the files in the created folders

import os
import tarfile

ham_folder = "/content/ham_examples"
spam_folder = "/content/spam_examples"

# List only tar files in ham folder
ham_examples = [
    os.path.join(ham_folder, f)
    for f in os.listdir(ham_folder)
    if tarfile.is_tarfile(os.path.join(ham_folder, f))
]

# List only tar files in spam folder
spam_examples = [
    os.path.join(spam_folder, f)
    for f in os.listdir(spam_folder)
    if tarfile.is_tarfile(os.path.join(spam_folder, f))
]


# Decompress ham tar files
for example in ham_examples:
    with tarfile.open(example) as tar:
        tar.extractall(path=ham_folder)
    os.remove(example)


# Decompress spam tar files
for example in spam_examples:
    with tarfile.open(example) as tar:
        tar.extractall(path=spam_folder)
    os.remove(example)


/tmp/ipython-input-1860702733.py:27: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tar.extractall(path=ham_folder)
/tmp/ipython-input-1860702733.py:34: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tar.extractall(path=spam_folder)


# Step 2 : Understanding data and familiarize with the data format

In [3]:
import os
import tarfile

#Define the folder paths
easy_ham_folder = "/content/ham_examples/easy_ham"
easy_ham_folder_2 = "/content/ham_examples/easy_ham_2"
hard_ham_folder = "/content/ham_examples/hard_ham"
spam_folder = "/content/spam_examples/spam"
spam_2_folder = "/content/spam_examples/spam_2"

#Define the document path for each folder
easy_ham_examples = [
    os.path.join(easy_ham_folder, f)
    for f in os.listdir(easy_ham_folder)
]

easy_ham_examples_2 = [
    os.path.join(easy_ham_folder_2,f) for f in os.listdir(easy_ham_folder_2)
]


hard_ham_examples = [
    os.path.join(hard_ham_folder, f) for f in os.listdir(hard_ham_folder)
]


spam_examples = [
    os.path.join(spam_folder, f) for f in os.listdir(spam_folder)
]

spam_2_examples = [
    os.path.join(spam_2_folder, f) for f in os.listdir(spam_2_folder)
]


#Define a sample number to take from a folder
sample_number = 1




In [4]:
#Easy ham example
example = easy_ham_examples[sample_number]
with open(example, "r", encoding= 'latin-1') as f:
    print(f.read())

From rssfeeds@example.com  Tue Oct  8 10:55:48 2002
Return-Path: <rssfeeds@example.com>
Delivered-To: zzzz@localhost.example.com
Received: from localhost (jalapeno [127.0.0.1])
	by example.com (Postfix) with ESMTP id 180F216F16
	for <zzzz@localhost>; Tue,  8 Oct 2002 10:55:48 +0100 (IST)
Received: from jalapeno [127.0.0.1]
	by localhost with IMAP (fetchmail-5.9.0)
	for zzzz@localhost (single-drop); Tue, 08 Oct 2002 10:55:48 +0100 (IST)
Received: from dogma.slashnull.org (localhost [127.0.0.1]) by
    dogma.slashnull.org (8.11.6/8.11.6) with ESMTP id g9880TK06063 for
    <zzzz@example.com>; Tue, 8 Oct 2002 09:00:29 +0100
Message-Id: <200210080800.g9880TK06063@dogma.slashnull.org>
To: zzzz@example.com
From: boingboing <rssfeeds@example.com>
Subject: Steven Levy's wireless neighbors
Date: Tue, 08 Oct 2002 08:00:29 -0000
Content-Type: text/plain; encoding=utf-8

URL: http://boingboing.net/#85534626
Date: Not supplied

After discovering an open wireless net available from his sofa, Steven 


In [5]:
#Easy ham example 2
example = easy_ham_examples_2[sample_number]
with open(example, "r", encoding= 'latin-1') as f:
    print(f.read())

From fork-admin@xent.com  Tue Aug  6 11:59:27 2002
Return-Path: <fork-admin@xent.com>
Delivered-To: yyyy@localhost.netnoteinc.com
Received: from localhost (localhost [127.0.0.1])
	by phobos.labs.netnoteinc.com (Postfix) with ESMTP id 88B02441F9
	for <jm@localhost>; Tue,  6 Aug 2002 06:49:15 -0400 (EDT)
Received: from phobos [127.0.0.1]
	by localhost with IMAP (fetchmail-5.9.0)
	for jm@localhost (single-drop); Tue, 06 Aug 2002 11:49:15 +0100 (IST)
Received: from xent.com ([64.161.22.236]) by dogma.slashnull.org
    (8.11.6/8.11.6) with ESMTP id g74K5sv07715 for <jm@jmason.org>;
    Sun, 4 Aug 2002 21:05:55 +0100
Received: from lair.xent.com (localhost [127.0.0.1]) by xent.com (Postfix)
    with ESMTP id D7C8A29409F; Sun,  4 Aug 2002 13:03:03 -0700 (PDT)
Delivered-To: fork@spamassassin.taint.org
Received: from hotmail.com (dav70.law15.hotmail.com [64.4.22.205]) by
    xent.com (Postfix) with ESMTP id 9853C294099 for <fork@xent.com>;
    Sun,  4 Aug 2002 13:02:52 -0700 (PDT)
Received: fro

In [6]:
# Hard Ham example
example = hard_ham_examples[sample_number]
with open(example, "r", encoding= 'latin-1') as f:
    print(f.read())


Return-Path: <replies@oracleeblast.com>
Received: (qmail 19678 invoked by alias); 10 Jul 2002 13:22:47 -0000
Received: (qmail 19416 invoked by uid 82); 10 Jul 2002 13:22:42 -0000
Received: from replies@oracleeblast.com by mailhost with qmail-scanner-1.00 (uvscan: v4.1.40/v4210. . Clean. Processed in 8.59332 secs); 10 Jul 2002 13:22:42 -0000
Received: from inet-mail6.oracle.com (209.246.10.170)
  by mi-1.rz.ruhr-uni-bochum.de with SMTP; 10 Jul 2002 13:22:30 -0000
Received: from blaster-smtp.oracle.com (eblast01.oracleeblast.com [148.87.9.11])
	by inet-mail6.oracle.com (Switch-2.2.2/Switch-2.2.0) with ESMTP id g6ADMHs25188
	for XXXXXX.YYYYY@RUHR-UNI-BOCHUM.DE; Wed, 10 Jul 2002 06:22:17 -0700 (PDT)
Date: Wed, 10 Jul 2002 06:22:17 -0700 (PDT)
Message-Id: <200207101322.g6ADMHs25188@inet-mail6.oracle.com>
Subject: Oracle Technology Network TechBlast - July 2002
From: Oracle Technology Network<replies@oracleeblast.com>
To: XXXXXX.YYYYY@RUHR-UNI-BOCHUM.DE
Reply-To: replies@oracleeblast.com
Con

In [7]:
#Spam example
example = spam_examples[sample_number]
with open(example, "r", encoding= 'latin-1') as f:
    print(f.read())

From spamassassin-talk-admin@lists.sourceforge.net  Wed Sep 11 13:57:59 2002
Return-Path: <spamassassin-talk-admin@example.sourceforge.net>
Delivered-To: zzzz@localhost.spamassassin.taint.org
Received: from localhost (jalapeno [127.0.0.1])
	by zzzzason.org (Postfix) with ESMTP id 36DC516F03
	for <zzzz@localhost>; Wed, 11 Sep 2002 13:57:58 +0100 (IST)
Received: from jalapeno [127.0.0.1]
	by localhost with IMAP (fetchmail-5.9.0)
	for zzzz@localhost (single-drop); Wed, 11 Sep 2002 13:57:58 +0100 (IST)
Received: from usw-sf-list2.sourceforge.net (usw-sf-fw2.sourceforge.net
    [216.136.171.252]) by dogma.slashnull.org (8.11.6/8.11.6) with ESMTP id
    g8BCnOC15167 for <zzzz-sa@jmason.org>; Wed, 11 Sep 2002 13:49:24 +0100
Received: from usw-sf-list1-b.sourceforge.net ([10.3.1.13]
    helo=usw-sf-list1.sourceforge.net) by usw-sf-list2.sourceforge.net with
    esmtp (Exim 3.31-VA-mm2 #1 (Debian)) id 17p6ua-00009Y-00; Wed,
    11 Sep 2002 05:48:04 -0700
Received: from panoramix.vasoftware.com 

In [8]:
#Spam example 2
example = spam_2_examples[sample_number]
with open(example, "r", encoding= 'latin-1') as f:
    print(f.read())


From 102646c1737aaa002@email.com  Mon Jun 24 17:06:36 2002
Return-Path: 102646c1737aaa002@email.com
Delivery-Date: Sat May 25 16:44:32 2002
Received: from mandark.labs.netnoteinc.com ([213.105.180.140]) by
    dogma.slashnull.org (8.11.6/8.11.6) with ESMTP id g4PFiVe03466 for
    <jm@jmason.org>; Sat, 25 May 2002 16:44:31 +0100
Received: from sercose01.sercoseseguros.com.br ([200.223.153.82]) by
    mandark.labs.netnoteinc.com (8.11.2/8.11.2) with ESMTP id g4PFiT713063 for
    <jm@netnoteinc.com>; Sat, 25 May 2002 16:44:30 +0100
Received: from mx1.mail.yahoo.com (wiley-1-342693.roadrunner.nf.net
    [205.251.195.54]) by sercose01.sercoseseguros.com.br with SMTP (Microsoft
    Exchange Internet Mail Service Version 5.5.2653.13) id LRRSDZ6N;
    Fri, 24 May 2002 11:33:08 -0300
Message-Id: <00005e38152b$00001e89$0000326a@email-com.mr.outblaze.com>
To: <Undisclosed.Recipients@mandark.labs.netnoteinc.com>
From: 102646c1737aaa002@email.com
Subject: Younger and healthier with ULTIMATE-hGH1728

# Step 3 : Integrate the downloaded ham and spam examples to construct a workable dataset

In [9]:
#Number of ham examples
total_ham_number = len(os.listdir(easy_ham_folder)) + len(os.listdir(easy_ham_folder_2)) + len(os.listdir(hard_ham_folder))

#Total number of spam

total_spam_number = len(os.listdir(spam_folder)) + len(os.listdir(spam_2_folder))

#Complete dataset size
total_dataset_size = total_ham_number + total_spam_number

#Proportion of ham
ham_proportion = total_ham_number / total_dataset_size

#proportion of spam
spam_proportion = total_spam_number / total_dataset_size



print(f"Total number of ham examples: {total_ham_number}")
print(f"Total number of spam examples: {total_spam_number}")
print(f'Total number of regiters of the dataset: {total_dataset_size}')
print(f'Proportion of ham: {ham_proportion}')
print(f'Proportion of spam: {spam_proportion}')






Total number of ham examples: 6954
Total number of spam examples: 2400
Total number of regiters of the dataset: 9354
Proportion of ham: 0.7434252726106478
Proportion of spam: 0.25657472738935216


In [10]:
#Construct the data from the data in the email
import os
import tarfile
import pandas as pd

#Ham dataset
total_ham_examples = easy_ham_examples + easy_ham_examples_2 + hard_ham_examples
#Spam dataset
total_spam_examples = spam_examples + spam_2_examples

#Pass from plain text ham into ham dataset
ham_rows = []
for example in total_ham_examples:
  with open(example,'r',encoding = 'latin-1') as f:
      ham_text = f.read()
  row = ham_rows.append({'message':ham_text,'label':0})


#Ham data set
df_ham = pd.DataFrame(ham_rows)


#Pass from plain text spam into spam dataset
spam_rows = []
for example in total_spam_examples:
  with open(example, 'r', encoding = 'latin-1') as f:
    spam_text = f.read()
  row = spam_rows.append({'message':spam_text,'label':1})


#Spam data set
df_spam = pd.DataFrame(spam_rows)


#Spam and ham Data set

X = pd.concat([df_ham,df_spam])

y = X['label']


# Step 4 : Hands on with ML model construction

In [11]:
#Import the necessary librarys for the project

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import  roc_auc_score
from sklearn.metrics import roc_curve
from sklearn.metrics import precision_recall_curve
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import SGDClassifier






#Step 5 : Splitting the training dataset into Train and Test sets.

In [12]:
#Staring with the strafied sampling
split  = StratifiedShuffleSplit(n_splits = 1, test_size = 0.2, train_size = 0.8, random_state = 42)

for train_idx,test_idx in split.split(X,y):
  X_train = X.iloc[train_idx]
  X_test = X.iloc[test_idx]
  y_train = y.iloc[train_idx]
  y_test = y.iloc[test_idx]


#Getting proportions of stratification in training set

train_proportion = y_train.value_counts() / len(y_train)

#Getting proportions of stratification in test set

test_proportion = y_test.value_counts() / len(y_test)


#Proportions in training set
print(f'Proportion of ham in the training set: {train_proportion[0]}')
print(f'Proportion of spam in the training set: {train_proportion[1]}')

#Proportions in test set

print(f'Proportion of ham in the test set: {test_proportion[0]}')
print(f'Proportion of spam in the test:{test_proportion[1]}')




Proportion of ham in the training set: 0.7434184150741681
Proportion of spam in the training set: 0.2565815849258319
Proportion of ham in the test set: 0.743452699091395
Proportion of spam in the test:0.256547300908605


# Step 6 : Creating custom transformers to deal with  data

In [13]:
#Transformer to deal with line breaks

from sklearn.base import BaseEstimator, TransformerMixin

class FormatText(BaseEstimator, TransformerMixin):
    def __init__(self,format= False):
        self.format = format

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        # X is a pandas Series of messages
        col = X.columns[0]
        cleaned =  X[col].apply(self._clean_message)
        return cleaned

    def _clean_message(self, msg):
        # Split and rejoin by line breaks
        lines = msg.split("\n")
        return "\n".join(lines)


In [14]:
#Transformer to deal with the Email headers
import email
from email import policy
import pandas as pd
from sklearn.base import BaseEstimator, TransformerMixin

class Header_off(BaseEstimator, TransformerMixin):
    def __init__(self, header_off=False):
        self.header_off = header_off

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        no_header = X.apply(self.remove_headers)
        return no_header  # FIX 1: preserve column name

    def remove_headers(self, msg_string):
        msg = email.message_from_string(msg_string, policy=policy.default)

        # Try simple payload first
        payload = msg.get_payload()

        # Case 1: simple email → string payload
        if isinstance(payload, str):
            return payload

        # Case 2: multipart → join plain text parts
        if isinstance(payload, list):
            parts = []
            for part in payload:
                try:
                    if part.get_content_type() == "text/plain":
                        parts.append(part.get_payload(decode=True).decode(errors="ignore"))
                except:
                    pass
            if parts:
                return "\n".join(parts)

        # Case 3: fallback — split headers from body
        raw = msg_string
        if "\n\n" in raw:
            return raw.split("\n\n", 1)[1]

        return raw  # last fallback


In [15]:
# Transformer to deal with URLS
import re
import pandas as pd
from sklearn.base import BaseEstimator, TransformerMixin

class URL_off(BaseEstimator, TransformerMixin):
    def __init__(self, url_off=False):
        self.url_off = url_off
        self.url_pattern = re.compile(r'https?://\S+|www\.\S+')

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        cleaned_series = X.apply(self.remove_urls)
        return cleaned_series

    def remove_urls(self, msg):
        return self.url_pattern.sub('', msg)


# Step 7 : Pipeline construction to deal with preprocessing and model fitting

In [16]:
#Create a preprocessing pipeline
preprocessing_pipeline = Pipeline(
    steps = [
        ('format_text', FormatText(True)),
        ('header_off', Header_off(True)),
        ('url_off', URL_off(True)),
        ('countvectorizer', CountVectorizer(stop_words='english')),
        ('tfidf', TfidfTransformer()),
        ('classifier', SGDClassifier(random_state =  42))




    ]
)



#Execute the pipeline with such classifier
preprocessing_pipeline.fit(X_train, y_train)


Pipeline(steps=[('format_text', FormatText(format=True)),
                ('header_off', Header_off(header_off=True)),
                ('url_off', URL_off(url_off=True)),
                ('countvectorizer', CountVectorizer(stop_words='english')),
                ('tfidf', TfidfTransformer()),
                ('classifier', SGDClassifier(random_state=42))])

# Step 8 : Evaluating model performance with classification metrics such as : f1 score, precision, recall, ROC AUC and accuracy.

# Comparison with dummy model


In [18]:
from sklearn.base import BaseEstimator, ClassifierMixin
import numpy as np
from scipy import stats

class DummyClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, strategy='most_frequent'):
        self.strategy = strategy
        self.most_frequent_class_ = None
        self.classes_ = None

    def fit(self, X, y):
        # 1. Identify all unique classes (e.g., [0, 1] or ['cat', 'dog'])
        self.classes_ = np.unique(y)

        # 2. Logic for 'most_frequent' strategy
        if self.strategy == 'most_frequent':
            # stats.mode returns the most common value
            # mode result is an object; .mode accesses the actual value
            mode_result = stats.mode(y, keepdims=True)
            self.most_frequent_class_ = mode_result.mode[0]

        return self

    def predict(self, X):
        # Check if fitted
        if self.most_frequent_class_ is None:
            raise Exception("Classifier not fitted yet!")

        # Return array of shape (n_samples,) filled with the majority class
        return np.full(shape=(X.shape[0],), fill_value=self.most_frequent_class_)

    def predict_proba(self, X):
        # Optional: Many sklearn metrics (like log_loss) need probabilities
        # This implementation simply gives 100% probability to the majority class
        if self.most_frequent_class_ is None:
            raise Exception("Classifier not fitted yet!")

        n_samples = X.shape[0]
        n_classes = len(self.classes_)

        # Create an array of zeros
        proba = np.zeros((n_samples, n_classes))

        # Find the index of the majority class
        majority_index = np.where(self.classes_ == self.most_frequent_class_)[0][0]

        # Set probability to 1.0 for the majority class
        proba[:, majority_index] = 1.0

        return proba

# Calculate dummy training set predictions

In [26]:
#Set the classifier
dummy_clf = DummyClassifier(strategy='most_frequent')
#Score of the dummy model
y_dummy_scores = cross_val_score(dummy_clf, X_train, y_train, cv=5, n_jobs=-1, scoring = 'accuracy')
#Predictions of the dummy model
y_dummy_pred = cross_val_predict(dummy_clf, X_train, y_train, cv=5, n_jobs = -1)
#Decision function of the model
y_dummy_proba = cross_val_predict(dummy_clf, X_train, y_train, cv=5, n_jobs= -1, method = 'predict_proba')


print(f'The accuracy score of the dummy moodel is : {np.mean(y_dummy_scores)}')




The accuracy score of the dummy moodel is : 0.7434165121687225


# Evaluate the classifier predictions in the training set

In [27]:
#Evaluate model in training set using k-fold validation

#Predictions of the model
y_pred = cross_val_predict(preprocessing_pipeline, X_train, y_train, cv=5,n_jobs = -1)

#Score of the model
y_scores = cross_val_score(preprocessing_pipeline, X_train, y_train, cv=5, scoring="accuracy", n_jobs=-1)

#Scores
y_proba_scores = cross_val_predict(preprocessing_pipeline, X_train, y_train, cv=5, n_jobs=-1, method="decision_function")


print(f'The accuracy score of the model is : {np.mean(y_scores)}')


The accuracy score of the model is : 0.9899770128492278


# Evaluation of confussion matrix of both models

# Confusion matrix of classifier

In [28]:
#Confusion matrix
confusion_matrix(y_train, y_pred)

array([[5548,   15],
       [  60, 1860]])

# Confusion matrix of dummy classifier

In [33]:
confusion_matrix(y_train, y_dummy_pred)

array([[5563,    0],
       [1920,    0]])

# Precision of model

In [41]:
#Precision
p = precision_score(y_train, y_pred)
print(f'The precision score of the model is : {p}')


The precision score of the model is : 0.992


# Precision of dummy model

In [42]:
q = precision_score(y_train, y_dummy_pred,zero_division = 0)
print(f'The precision score of the dummy model is : {q}')

The precision score of the dummy model is : 0.0


# Model recall

In [44]:
#Recall

r = recall_score(y_train, y_pred)
print(f'The recall score of the model is : {r}')

The recall score of the model is : 0.96875


# Dummy model recall

In [43]:
s = recall_score(y_train, y_dummy_pred,zero_division = 0)
print(f'The recall score of the dummy model is : {s}')

The recall score of the dummy model is : 0.0


# Model f1 score

In [46]:
#f1_score

f = f1_score(y_train, y_pred)
print(f'The f1 score of the model is : {f}')

The f1 score of the model is : 0.9802371541501976


# Dummy model f1 score

In [45]:
h = f1_score(y_train, y_dummy_pred,zero_division = 0)
print(f'The f1 score of the dummy model is : {h}')

The f1 score of the dummy model is : 0.0


# AUC of model

In [48]:
#AUC

a = roc_auc_score(y_train, y_proba_scores)
print(f'The AUC of the model is : {a}')

The AUC of the model is : 0.9992789973934927


# AUC of dummy model

In [47]:
b = roc_auc_score(y_train, y_dummy_proba[:,1])
print(f'The AUC of the dummy model is : {b}')

The AUC of the dummy model is : 0.5


Model Performance Comparison: SGD Classifier vs. Dummy Baseline


1. Executive Summary
The objective of this analysis was to validate the predictive power of the Stochastic Gradient Descent (SGD) Classifier by comparing it against a Dummy Classifier baseline. While the baseline represents a "naive" strategy (predicting only the majority class), the SGD model demonstrates near-perfect classification, effectively capturing the underlying patterns in the data.



3. Critical Analysis
 * The "Accuracy Trap" & Class ImbalanceThe Dummy Classifier achieved an accuracy of $74.34\%$ simply by exploiting the class imbalance (guessing the most frequent category). However, its F1-Score of $0.0$ reveals that it failed to identify a single positive case. This confirms that accuracy alone is a deceptive metric for this dataset and justifies the need for our advanced model.
 * Reliability and CoverageThe SGD Classifier achieved a Precision of $99.20\%$, meaning it is almost never wrong when it identifies a positive target. Combined with a Recall of $96.87\%$, the model proves it is both highly selective and highly thorough.
 * Discrimination Power (AUC)The ROC-AUC score of $0.999$ represents near-perfect separation. While the Dummy baseline is equivalent to a coin flip ($0.50$), our model can rank samples by probability with total confidence.
 * ConclusionThe comparison demonstrates that the SGD Classifier is an exceptionally robust tool. By reducing the error rate from $25.66\%$ (Dummy) to a mere $1.01\%$ (Model), we have developed a system that is significantly more reliable than frequency-based guessing. The model is statistically stable and ready for production deployment.



# Step 9 : Report test score to the managers with different classification metrics

In [49]:
#Score in test set using accuracy
test_pred = preprocessing_pipeline.predict(X_test)
test_score = accuracy_score(y_test, test_pred)
print(test_score)


0.9903794762159273


In [50]:
#Score in test set using f1_score
test_f1_score = f1_score(y_test, test_pred)
print(test_f1_score)

0.9811715481171548


In [51]:
y_test_scores = preprocessing_pipeline.decision_function(X_test)

In [52]:
#Score in test set using AUC
test_auc = roc_auc_score(y_test, y_test_scores)
print(test_auc)


0.9995431943445963
